In [94]:
import pandas as pd
import numpy as np

import json
import pickle
#import datetime
from datetime import datetime, date, timedelta

import warnings
warnings.filterwarnings('ignore')

In [95]:
#!pip freeze > requirements.txt 

In [96]:
#coord = pd.read_excel('data/terminal_data_hackathon v4.xlsx', sheet_name='TIDS', index_col=0)
#incomes = pd.read_excel('data/terminal_data_hackathon v4.xlsx', sheet_name='Incomes', index_col='TID')
#times = pd.read_csv('data/times v4.csv')

In [97]:
today = date.today()

In [98]:
with open('CloudRun5/results_2023-05-26_2.6_0.9_500000_0.pickle', 'rb') as f:
    cl= pd.read_pickle(f)

In [27]:
cl.bags # приоритетные точки по тз, в которые не заехали

{0: set(),
 1: set(),
 2: set(),
 3: set(),
 4: set(),
 5: set(),
 6: set(),
 7: set(),
 8: set(),
 9: set(),
 10: set(),
 11: set(),
 12: set(),
 13: set(),
 14: set(),
 15: set(),
 16: set(),
 17: set(),
 18: set(),
 19: set(),
 20: set(),
 21: set(),
 22: set(),
 23: set(),
 24: set(),
 25: set(),
 26: set(),
 27: set(),
 28: set(),
 29: set(),
 30: set(),
 31: set(),
 32: set(),
 33: set(),
 34: set(),
 35: set(),
 36: set(),
 37: set(),
 38: set(),
 39: set(),
 40: set(),
 41: set(),
 42: set(),
 43: set(),
 44: set(),
 45: set(),
 46: set(),
 47: set(),
 48: set(),
 49: set(),
 50: set(),
 51: set(),
 52: set(),
 53: set(),
 54: set(),
 55: set(),
 56: set(),
 57: set(),
 58: set(),
 59: set(),
 60: set(),
 61: set(),
 62: set(),
 63: set(),
 64: set(),
 65: set(),
 66: set(),
 67: set(),
 68: set(),
 69: set(),
 70: set(),
 71: set(),
 72: set(),
 73: set(),
 74: set(),
 75: set(),
 76: set(),
 77: set(),
 78: set(),
 79: set(),
 80: set(),
 81: set(),
 82: set(),
 83: set(),
 8

In [28]:
cl.balance # остатки

остаток на 31.08.2022 (входящий)  2022-09-01 00:00:00   
TID                                                             
406136                            160000               250000  \
406139                            387000               490000   
406145                            287000               430000   
406148                            355000                50000   
406180                            597000                96000   
...                                  ...                  ...   
699578                             47000                    0   
699579                            236000               285000   
699629                             67000                63000   
699641                            278000                63000   
699664                            333000               333000   

        2022-09-02 00:00:00  2022-09-03 00:00:00  2022-09-04 00:00:00   
TID                                                                     
406136               355000               454000               561000  \
406139               206000               374000               498000   
406145               136000               260000               377000   
406148               123000               176000               241000   
406180               178000               249000                72000   
...                     ...                  ...                  ...   
699578                    0                10000                    0   
699579               353000               353000                    0   
699629               126000               189000                64000   
699641                63000                63000                51000   
699664               333000               380000                    0   

        2022-09-05 00:00:00  2022-09-06 00:00:00  2022-09-07 00:00:00   
TID                                                                     
406136               110000               170000               245000  \
406139               576000               741000               905000   
406145               500000               640000               779000   
406148               316000               416000                53000   
406180               158000               213000               268000   
...                     ...                  ...                  ...   
699578                10000                16000                16000   
699579                13000                63000                63000   
699629               138000               138000               259000   
699641                51000                51000               128000   
699664                43000                43000                43000   

        2022-09-08 00:00:00  2022-09-09 00:00:00  ...  2022-11-21 00:00:00   
TID                                               ...                        
406136               334000               429000  ...               302000  \
406139               174000               262000  ...               776000   
406145               917000               118000  ...               526000   
406148               105000               187000  ...               574000   
406180               343000                75000  ...               215000   
...                     ...                  ...  ...                  ...   
699578                41000                    0  ...                98000   
699579               195000               230000  ...               426000   
699629               322000               394000  ...               611000   
699641               128000               168000  ...                94000   
699664                79000               114000  ...               257000   

        2022-11-22 00:00:00  2022-11-23 00:00:00  2022-11-24 00:00:00   
TID                                                                     
406136               380000               380000               545000  \
406139               929000 

In [67]:
cl.day_df_dict[2]

approx_time_cost_rub  outlier   pred_balance  priority   
TID                                                              
406136           -654.583333        0  354947.335051        -1  \
406139           -531.865079        0  682945.578202        -1   
406145           -606.289683        0  574857.018780        -1   
406148           -592.301587        0  127165.478087        -1   
406180           -498.710317        0  171373.577790        -1   
...                      ...      ...            ...       ...   
699578           -475.773810        0    9214.754100        -1   
699579           -530.158730        0  329577.049180        -1   
699629          -1059.404762        0   89750.561840        -1   
699641           -497.738095        0   65725.206444        -1   
699664           -461.150794        0  371058.646390        -1   

        cost_per_service  cash_rate_per_day    all_costs   
TID                                                        
406136            -100.0          19.449169  -735.134164  \
406139            -100.0          37.421676  -594.443404   
406145            -100.0          31.499015  -674.790668   
406148            -100.0           6.967971  -685.333616   
406180            -100.0           9.390333  -589.319984   
...                  ...                ...          ...   
699578            -100.0           0.504918  -575.268891   
699579            -100.0          18.059016  -612.099714   
699629            -100.0           4.917839 -1154.486923   
699641            -100.0           3.601381  -594.136714   
699664            -100.0          20.331981  -540.818813   

        costs_without_vehicle  pred_balance_no_trust_koef   
TID                                                         
406136             -80.550831               345406.668228  \
406139             -62.578324               665405.071092   
406145             -68.500985               561688.198891   
406148             -93.032029               120150.434624   
406180             -90.609667               164521.434355   
...                       ...                         ...   
699578             -99.495082                 8377.049181   
699579             -81.940984               325524.590163   
699629             -95.082161                87318.692582   
699641             -96.398619                65477.460404   
699664             -79.668019               367598.769445   

        cost_per_service_no_trust_koef  cash_rate_per_day_no_trust_koef   
TID                                                                       
406136                          -100.0                        18.926393  \
406139                          -100.0                        36.460552   
406145                          -100.0                        30.777436   
406148                          -100.0                         6.583585   
406180                          -100.0                         9.014873   
...                                ...                              ...   
699578                          -100.0                         0.459016   
699579                          -100.0                        17.836964   
699629                          -100.0                         4.784586   
699641                          -100.0                         3.587806   
699664                          -100.0                        20.142398   

        nes_degree  penalty  
TID                          
406136         2.0      752  
406139         2.0      968  
406145         2.0      885  
406148         1.0      107  
406180         1.0      110  
...            ...      ...  
699578         1.0      100  
699579         2.0      739  
699629         1.0      105  
699641         1.0      103  
699664         2.0      761  

[1630 rows x 13 columns]

In [30]:
cl.over_balance_limit # терминалы, которые всё же переполнились (прогноз прироста наличности был существенно отличен от факта)

{1: {674406},
 2: set(),
 3: {605329},
 4: set(),
 5: {676981},
 6: {664161},
 7: set(),
 8: {605329},
 9: {607519},
 10: set(),
 11: {682268},
 12: {608567},
 13: {616681},
 14: {605329},
 15: set(),
 16: set(),
 17: {622398},
 18: set(),
 19: {676383},
 20: {605329},
 21: {676981, 677522},
 22: {664161},
 23: {676981},
 24: {687408},
 25: set(),
 26: {658930},
 27: {605329, 616681},
 28: {676981},
 29: {606531},
 30: set(),
 31: {605329, 616480, 686130},
 32: {406509},
 33: set(),
 34: {664165, 676981},
 35: {616681},
 36: {640027},
 37: {657936},
 38: {635757},
 39: set(),
 40: set(),
 41: {622352, 676981},
 42: {406139, 616681},
 43: {677563},
 44: {605329, 683789, 688376},
 45: {606118, 607597, 609954, 636462, 682273},
 46: set(),
 47: {638490},
 48: {676981},
 49: set(),
 50: {605329, 644016},
 51: {686118},
 52: set(),
 53: set(),
 54: set(),
 55: {605329, 637522, 687408},
 56: set(),
 57: {634024},
 58: {606531, 683117},
 59: {634558, 683756, 692611},
 60: set(),
 61: {406504},

In [103]:
lst=[]
for i in cl.over_balance_limit.values():
    lst+=list(i)

In [108]:
def create_day_route_df(obj, day_num, vehicle):
    '''Вывод датасета по маршруту одного броневика.
    Входные параметры:
    obj: класс отработавшего оптимизатора
    day_num: номер дня от 1 до 91
    vehicle: броневик, значения из списка [0,1,2,3,4] '''
    
    y=[x+1 for x in obj.route_vehicle[day_num][vehicle]]
    dist_with_service = [int(obj.distance_matrix[y[i]][y[i+1]]) for i in range(len(y)-1)]
    dist= list(map(lambda x: max(0, x-10), dist_with_service))
    
    day_route_df = pd.DataFrame(index=obj.balance.iloc[obj.route_vehicle[day_num][vehicle][1:]].index,
                            columns=['vehicle_num','tid','start_time', 'end_time', 'route_time_to_tid',
                                     'route_time_to_tid_with_service'])
    day_route_df['vehicle_num'] = vehicle
    day_route_df['tid'] = day_route_df.index
    start_day = pd.to_datetime(obj.balance.iloc[:,day_num].name)
    day_route_df.loc[:,'route_time_to_tid'] = dist
    day_route_df.loc[:,'route_time_to_tid_with_service'] = dist_with_service
    day_route_df['route_time_to_tid_with_service_cumsum'] = day_route_df.route_time_to_tid_with_service.cumsum()
    day_route_df.loc[:,'start_time'] = start_day + day_route_df.route_time_to_tid_with_service_cumsum.apply(lambda x: 
                                            timedelta(hours=8, minutes=x))
    day_route_df['end_time'] = day_route_df['start_time'] + timedelta(minutes=obj.service_time)
    return day_route_df

In [109]:
create_day_route_df(cl, 90, 0)
# расстояния в минутах округлены для удобства в бОльшую сторону (12.1 мин => 13 мин) 

vehicle_num     tid           start_time             end_time   
TID                                                                     
657216            0  657216  2022-11-29 08:00:00  2022-11-29 08:10:00  \
698718            0  698718  2022-11-29 08:12:00  2022-11-29 08:22:00   
681975            0  681975  2022-11-29 08:30:00  2022-11-29 08:40:00   
649851            0  649851  2022-11-29 08:45:00  2022-11-29 08:55:00   
636409            0  636409  2022-11-29 09:03:00  2022-11-29 09:13:00   
683789            0  683789  2022-11-29 09:18:00  2022-11-29 09:28:00   
689149            0  689149  2022-11-29 09:33:00  2022-11-29 09:43:00   
698710            0  698710  2022-11-29 09:48:00  2022-11-29 09:58:00   
682752            0  682752  2022-11-29 10:03:00  2022-11-29 10:13:00   
629749            0  629749  2022-11-29 10:25:00  2022-11-29 10:35:00   
616681            0  616681  2022-11-29 10:38:00  2022-11-29 10:48:00   
668401            0  668401  2022-11-29 10:55:00  2022-11-29 11:05:00   
653690            0  653690  2022-11-29 11:22:00  2022-11-29 11:32:00   
669434            0  669434  2022-11-29 11:41:00  2022-11-29 11:51:00   
606128            0  606128  2022-11-29 11:52:00  2022-11-29 12:02:00   
636077            0  636077  2022-11-29 12:07:00  2022-11-29 12:17:00   
659661            0  659661  2022-11-29 12:20:00  2022-11-29 12:30:00   
407025            0  407025  2022-11-29 12:40:00  2022-11-29 12:50:00   
686161            0  686161  2022-11-29 12:55:00  2022-11-29 13:05:00   
687662            0  687662  2022-11-29 13:18:00  2022-11-29 13:28:00   
607599            0  607599  2022-11-29 13:43:00  2022-11-29 13:53:00   
622379            0  622379  2022-11-29 13:54:00  2022-11-29 14:04:00   
608061            0  608061  2022-11-29 14:24:00  2022-11-29 14:34:00   
608071            0  608071  2022-11-29 14:38:00  2022-11-29 14:48:00   
605329            0  605329  2022-11-29 14:52:00  2022-11-29 15:02:00   
619514            0  619514  2022-11-29 15:19:00  2022-11-29 15:29:00   
634656            0  634656  2022-11-29 15:33:00  2022-11-29 15:43:00   
629701            0  629701  2022-11-29 15:48:00  2022-11-29 15:58:00   
606001            0  606001  2022-11-29 16:03:00  2022-11-29 16:13:00   
637527            0  637527  2022-11-29 16:18:00  2022-11-29 16:28:00   
676359            0  676359  2022-11-29 16:40:00  2022-11-29 16:50:00   
627726            0  627726  2022-11-29 17:09:00  2022-11-29 17:19:00   
622450            0  622450  2022-11-29 17:23:00  2022-11-29 17:33:00   
698628            0  698628  2022-11-29 17:38:00  2022-11-29 17:48:00   
607545            0  607545  2022-11-29 18:00:00  2022-11-29 18:10:00   
622373            0  622373  2022-11-29 18:12:00  2022-11-29 18:22:00   
607542            0  607542  2022-11-29 18:30:00  2022-11-29 18:40:00   
644568            0  644568  2022-11-29 19:29:00  2022-11-29 19:39:00   
608007            0  608007  2022-11-29 19:46:00  2022-11-29 19:56:00   

       route_time_to_tid route_time_to_tid_with_service   
TID                                                       
657216                 0                              0  \
698718                 2                             12   
681975                 8                             18   
649851                 5                             15   
636409                 8                             18   
683789                 5                             15   
689149                 5                             15   
698710                 5                             15   
682752                 5                             15   
629749                12                             22   
616681                 3                             13   
668401                 7                             17   
653690                17                             27   
669434                 9                             19   
606128                 1                             11   
636077     

In [110]:
%%time
report_route = pd.DataFrame()
for d in range(1, 92):
    for v in range(5):
        report_route = pd.concat([report_route, create_day_route_df(cl, d, v)])

Wall time: 2.51 s


In [111]:
report_route

vehicle_num     tid           start_time             end_time   
TID                                                                     
677522            0  677522  2022-09-01 08:00:00  2022-09-01 08:10:00  \
677559            0  677559  2022-09-01 08:36:00  2022-09-01 08:46:00   
634652            0  634652  2022-09-01 08:53:00  2022-09-01 09:03:00   
634703            0  634703  2022-09-01 09:34:00  2022-09-01 09:44:00   
688325            0  688325  2022-09-01 10:04:00  2022-09-01 10:14:00   
...             ...     ...                  ...                  ...   
665569            4  665569  2022-11-30 18:18:00  2022-11-30 18:28:00   
622276            4  622276  2022-11-30 18:35:00  2022-11-30 18:45:00   
639707            4  639707  2022-11-30 18:48:00  2022-11-30 18:58:00   
619201            4  619201  2022-11-30 19:32:00  2022-11-30 19:42:00   
606168            4  606168  2022-11-30 19:49:00  2022-11-30 19:59:00   

       route_time_to_tid route_time_to_tid_with_service   
TID                                                       
677522                 0                              0  \
677559                26                             36   
634652                 7                             17   
634703                31                             41   
688325                20                             30   
...                  ...                            ...   
665569                 8                             18   
622276                 7                             17   
639707                 3                             13   
619201                34                             44   
606168                 7                             17   

       route_time_to_tid_with_service_cumsum  
TID                                           
677522                                     0  
677559                                    36  
634652                                    53  
634703                                    94  
688325                                   124  
...                                      ...  
665569                                   618  
622276                                   635  
639707                                   648  
619201                                   692  
606168                                   709  

[14299 rows x 7 columns]

In [18]:
cl.balance

остаток на 31.08.2022 (входящий)  2022-09-01 00:00:00   
TID                                                             
406136                            160000             259000.0  \
406139                            387000             500300.0   
406145                            287000             444300.0   
406148                            355000              50000.0   
406180                            597000              96000.0   
...                                  ...                  ...   
699578                             47000                  0.0   
699579                            236000             289900.0   
699629                             67000              63000.0   
699641                            278000              63000.0   
699664                            333000             333000.0   

        2022-09-02 00:00:00  2022-09-03 00:00:00  2022-09-04 00:00:00   
TID                                                                     
406136             374500.0             483400.0             601100.0  \
406139             206000.0             390800.0             527200.0   
406145             593900.0             124000.0             252700.0   
406148             130300.0             188600.0             260100.0   
406180             186200.0             264300.0             343500.0   
...                     ...                  ...                  ...   
699578                  0.0              10000.0              10000.0   
699579             364700.0                  0.0                  0.0   
699629             132300.0             201600.0             272000.0   
699641              63000.0              63000.0             119100.0   
699664             333000.0             384700.0             384700.0   

        2022-09-05 00:00:00  2022-09-06 00:00:00  2022-09-07 00:00:00   
TID                                                                     
406136             110000.0             176000.0             258500.0  \
406139             613000.0             794500.0             974900.0   
406145             388000.0             542000.0             694900.0   
406148             342600.0             100000.0             158300.0   
406180             438100.0              55000.0             115500.0   
...                     ...                  ...                  ...   
699578              21000.0               6000.0               6000.0   
699579              14300.0              50000.0              50000.0   
699629             353400.0                  0.0             133100.0   
699641             119100.0                  0.0              84700.0   
699664             432000.0                  0.0                  0.0   

        2022-09-08 00:00:00  2022-09-09 00:00:00  ...  2022-11-21 00:00:00   
TID                                               ...                        
406136             356400.0             460900.0  ...             754500.0  \
406139             174000.0             270800.0  ...             839600.0   
406145             846700.0             976500.0  ...             745600.0   
406148             215500.0             305700.0  ...             119800.0   
406180             198000.0             280500.0  ...             480600.0   
...                     ...                  ...  ...                  ...   
699578              33500.0              33500.0  ...              91300.0   
699579             195200.0             233700.0  ...             342600.0   
699629             202400.0             281600.0  ...             348400.0   
699641              84700.0             128700.0  ...             183700.0   
699664              39600.0              78100.0  ...              48800.0   

        2022-11-22 00:00:00  2022-11-23 00:00:00  2022-11-24 00:00:00   
TID                                                                     
406136             840300.0             840300.0             165000.0  \
406139             153000.0 

In [117]:
cost_per_service_fact_df = pd.DataFrame(index=cl.balance.index, columns=cl.balance.columns[1:])
for d in range(1,92):
    cost_per_service_fact_df.iloc[:,d-1] = cl.day_df_dict[d].cost_per_service_fact

In [123]:
cash_rate_per_day_fact_df = pd.DataFrame(index=cl.balance.index, columns=cl.balance.columns[1:])
for d in range(1,92):
    cash_rate_per_day_fact_df.iloc[:,d-1] = cl.day_df_dict[d].cash_rate_per_day_fact

In [124]:
cash_rate_per_day_fact_df

2022-09-01 00:00:00 2022-09-02 00:00:00 2022-09-03 00:00:00   
TID                                                                  
406136           -13.69863          -19.452055          -24.876712  \
406139          -26.849315                 0.0          -20.493151   
406145          -23.561644                 0.0          -14.246575   
406148                 0.0           -6.739726           -9.643836   
406180                 0.0           -9.753425          -13.643836   
...                    ...                 ...                 ...   
699578                 0.0                 0.0           -0.547945   
699579          -15.616438          -19.342466          -19.342466   
699629                 0.0            -6.90411          -10.356164   
699641                 0.0           -3.452055           -3.452055   
699664          -18.246575          -18.246575          -20.821918   

       2022-09-04 00:00:00 2022-09-05 00:00:00 2022-09-06 00:00:00   
TID                                                                  
406136          -30.739726                 0.0           -9.315068  \
406139          -27.287671          -31.561644           -40.60274   
406145          -20.657534           -27.39726          -35.068493   
406148                 0.0           -7.671233          -13.150685   
406180                 0.0           -8.657534          -11.671233   
...                    ...                 ...                 ...   
699578                 0.0           -0.547945           -0.876712   
699579                 0.0           -0.712329           -3.452055   
699629                 0.0           -7.561644           -7.561644   
699641                 0.0           -2.794521           -2.794521   
699664                 0.0           -2.356164           -2.356164   

       2022-09-07 00:00:00 2022-09-08 00:00:00 2022-09-09 00:00:00   
TID                                                                  
406136          -13.424658           -18.30137          -23.506849  \
406139          -49.589041                 0.0          -14.356164   
406145          -42.684932          -50.246575                 0.0   
406148          -16.054795           -18.90411           -23.39726   
406180          -14.684932          -18.794521           -22.90411   
...                    ...                 ...                 ...   
699578           -0.876712           -2.246575           -2.246575   
699579           -3.452055          -10.684932           -12.60274   
699629          -14.191781          -17.643836          -21.589041   
699641           -7.013699           -7.013699           -9.205479   
699664           -2.356164           -4.328767           -6.246575   

       2022-09-10 00:00:00  ... 2022-11-21 00:00:00 2022-11-22 00:00:00   
TID                         ...                                           
406136          -29.863014  ...          -29.424658           -33.69863  \
406139          -27.506849  ...          -42.520548           -50.90411   
406145          -13.369863  ...          -23.342466          -28.821918   
406148          -26.849315  ...           -38.30137                 0.0   
406180           -26.90411  ...               -32.0          -35.068493   
...                    ...  ...                 ...                 ...   
699578                 0.0  ...           -4.547945            -5.09589   
699579           -14.90411  ...          -17.534247          -21.260274   
699629          -24.712329  ...          -11.890411          -14.575342   
699641                 0.0  ...           -9.150685          -11.780822   
699664           -6.246575  ...           -9.534247          -10.520548   

       2022-11-23 00:00:00 2022-11-24 00:00:00 2022-11-25 00:00:00   
TID                                                                  
406136           -33.69863          -42.739726          -42.739726  \
406139                 0.0          -16.876712          -28.164384   
406145          -3

__Условия расчета остатка по ТЗ (как считали мы вначале):__

    2.  Принимаем, что остаток в терминале увеличивается на сумму оборота в терминале разово. Т.е. утром известны остатки в терминалах и они не меняются за время работы бригад броневиков. Остатки на утро = остатки утра вчера  + обороты вчера.
Например, для терминала 406136 на утро 02.09 остаток = 250000, если 01.09 терминал не обслужили и = 90000, если терминал 01.09 обслужили.


__Условия расчета остатка по Отчету (пересчитаем сейчас):__

    Лист "остатки на конец дня" показывает сумму остатков в устройствах в разрезе дат на конец дня. Т.е. в случае, если устройство было инкассировано, в ячейке точно должен быть 0

In [194]:
cor_balance=pd.DataFrame(data=0, index=cl.incomes.index, columns = cl.incomes.columns[1:])
cor_balance.insert(loc=0, column='остаток на 31.08.2022 (входящий)',
                            value=cl.incomes['остаток на 31.08.2022 (входящий)'])

In [195]:
cash_rate_per_day_fact_df_cor = cor_balance.iloc[:,1:]

In [196]:
%%time
d=1
while d < 92:
    no_visits = cl.day_df_dict[d].cost_per_service_fact.apply(lambda x: False if x < 0 else True)
    cor_balance.iloc[:,d][no_visits] = cor_balance.loc[no_visits].iloc[:,d-1] + cl.incomes.loc[no_visits].iloc[:,d]
    cash_rate_per_day_fact_df_cor.iloc[:,d-1][no_visits] = - cor_balance.loc[no_visits].iloc[:,d]*cl.p_per_day
    d+=1
    

Wall time: 637 ms


In [197]:
cor_balance

остаток на 31.08.2022 (входящий)  2022-09-01 00:00:00   
TID                                                             
406136                            160000               250000  \
406139                            387000               490000   
406145                            287000               430000   
406148                            355000                    0   
406180                            597000                    0   
...                                  ...                  ...   
699578                             47000                    0   
699579                            236000               285000   
699629                             67000                    0   
699641                            278000                    0   
699664                            333000               333000   

        2022-09-02 00:00:00  2022-09-03 00:00:00  2022-09-04 00:00:00   
TID                                                                     
406136               355000               454000               561000  \
406139                    0               168000               292000   
406145                    0               124000               241000   
406148                73000               126000                    0   
406180                82000               153000                    0   
...                     ...                  ...                  ...   
699578                    0                10000                    0   
699579               353000               353000                    0   
699629                63000               126000                    0   
699641                    0                    0                    0   
699664               333000               380000                    0   

        2022-09-05 00:00:00  2022-09-06 00:00:00  2022-09-07 00:00:00   
TID                                                                     
406136                    0                60000               135000  \
406139               370000               535000               699000   
406145               364000               504000               643000   
406148                75000               175000               228000   
406180                86000               141000               196000   
...                     ...                  ...                  ...   
699578                10000                16000                16000   
699579                13000                63000                63000   
699629                74000                74000               195000   
699641                    0                    0                77000   
699664                43000                43000                43000   

        2022-09-08 00:00:00  2022-09-09 00:00:00  ...  2022-11-21 00:00:00   
TID                                               ...                        
406136               224000               319000  ...               480000  \
406139                    0                88000  ...               636000   
406145               781000                    0  ...               353000   
406148               280000               362000  ...               574000   
406180               271000               346000  ...               507000   
...                     ...                  ...  ...                  ...   
699578                41000                41000  ...                83000   
699579               195000               230000  ...               226000   
699629               258000               330000  ...               166000   
699641                77000               117000  ...               167000   
699664                79000               114000  ...               174000   

        2022-11-22 00:00:00  2022-11-23 00:00:00  2022-11-24 00:00:00   
TID                                                                     
406136               558000               558000               723000  \
406139               789000 

In [198]:
cash_rate_per_day_fact_df_cor

2022-09-01 00:00:00  2022-09-02 00:00:00  2022-09-03 00:00:00   
TID                                                                     
406136           -13.698630           -19.452055           -24.876712  \
406139           -26.849315             0.000000            -9.205479   
406145           -23.561644             0.000000            -6.794521   
406148             0.000000            -4.000000            -6.904110   
406180             0.000000            -4.493151            -8.383562   
...                     ...                  ...                  ...   
699578             0.000000             0.000000            -0.547945   
699579           -15.616438           -19.342466           -19.342466   
699629             0.000000            -3.452055            -6.904110   
699641             0.000000             0.000000             0.000000   
699664           -18.246575           -18.246575           -20.821918   

        2022-09-04 00:00:00  2022-09-05 00:00:00  2022-09-06 00:00:00   
TID                                                                     
406136           -30.739726             0.000000            -3.287671  \
406139           -16.000000           -20.273973           -29.315068   
406145           -13.205479           -19.945205           -27.616438   
406148             0.000000            -4.109589            -9.589041   
406180             0.000000            -4.712329            -7.726027   
...                     ...                  ...                  ...   
699578             0.000000            -0.547945            -0.876712   
699579             0.000000            -0.712329            -3.452055   
699629             0.000000            -4.054795            -4.054795   
699641             0.000000             0.000000             0.000000   
699664             0.000000            -2.356164            -2.356164   

        2022-09-07 00:00:00  2022-09-08 00:00:00  2022-09-09 00:00:00   
TID                                                                     
406136            -7.397260           -12.273973           -17.479452  \
406139           -38.301370             0.000000            -4.821918   
406145           -35.232877           -42.794521             0.000000   
406148           -12.493151           -15.342466           -19.835616   
406180           -10.739726           -14.849315           -18.958904   
...                     ...                  ...                  ...   
699578            -0.876712            -2.246575            -2.246575   
699579            -3.452055           -10.684932           -12.602740   
699629           -10.684932           -14.136986           -18.082192   
699641            -4.219178            -4.219178            -6.410959   
699664            -2.356164            -4.328767            -6.246575   

        2022-09-10 00:00:00  ...  2022-11-21 00:00:00  2022-11-22 00:00:00   
TID                          ...                                             
406136           -23.835616  ...           -26.301370           -30.575342  \
406139           -17.972603  ...           -34.849315           -43.232877   
406145            -6.904110  ...           -19.342466           -24.821918   
406148           -23.287671  ...           -31.452055             0.000000   
406180           -22.958904  ...           -27.780822           -30.849315   
...                     ...  ...                  ...                  ...   
699578             0.000000  ...            -4.547945            -5.095890   
699579           -14.904110  ...           -12.383562           -16.109589   
699629           -21.205479  ...            -9.095890           -11.780822   
699641             0.000000  ...            -9.150685           -11.780822   
699664            -6.246575  ...            -9.534247           -10.520548   

        2022-11-23 00:00:00  2022-11-24 00:00:00  2022-11-25 00:00:00   
TID                                                                   

In [206]:
report_route_format = report_route.copy()
report_route_format['start_time'] = report_route_format['start_time'].apply(str)
report_route_format['end_time'] = report_route_format['end_time'].apply(str)

In [209]:
with pd.ExcelWriter("report.xlsx") as writer:
   
    cl.balance.to_excel(writer, sheet_name="остатки на конец дня как в тз")
    cash_rate_per_day_fact_df.to_excel(writer, sheet_name="стоимость фондирования как в тз")
    cor_balance.to_excel(writer, sheet_name="остатки на конец дня как в пояснениях к отчету")
    cash_rate_per_day_fact_df_cor.to_excel(writer, sheet_name="стоимость фондирования как в пояснениях к отчету")
    cost_per_service_fact_df.to_excel(writer, sheet_name="стоимость инкассации")
    report_route_format.iloc[:,:-1].to_excel(writer, sheet_name="маршруты", index=False)

In [218]:
cl.balance.to_csv('data/balance_all.csv')

In [219]:
pd.DataFrame(cl.balance.iloc[:,-1]).to_csv('data/balance_2022-11-30.csv')

In [236]:
terms_dict_by_date = {}
start_date = pd.to_datetime('2022-09-01 00:00:00')
for i in cl.terms_dict:
    terms_dict_by_date[str(start_date + timedelta(days=i-1))] = list(cl.terms_dict[i])

In [237]:
terms_dict_by_date

{'2022-08-31 00:00:00': [638981,
  688211,
  688213,
  688214,
  688219,
  688223,
  688225,
  688226,
  688241,
  688242,
  688247,
  688248,
  688254,
  688255,
  688262,
  688263,
  688267,
  688274,
  688279,
  688283,
  663710,
  688286,
  688293,
  688298,
  688301,
  688304,
  688315,
  688321,
  688323,
  688325,
  688326,
  688338,
  688339,
  688343,
  688345,
  688349,
  688351,
  688352,
  688355,
  688361,
  688365,
  688367,
  688376,
  688383,
  688389,
  688396,
  688397,
  606501,
  606506,
  606509,
  606511,
  606514,
  606515,
  606516,
  606518,
  606520,
  606521,
  606524,
  672061,
  606531,
  606533,
  606536,
  606537,
  606540,
  606541,
  606542,
  606545,
  606546,
  606553,
  606554,
  606556,
  606559,
  606562,
  606563,
  606565,
  606569,
  606572,
  606574,
  606575,
  606576,
  606578,
  606579,
  606584,
  680312,
  680315,
  680316,
  680319,
  680322,
  680324,
  606597,
  606598,
  680328,
  672190,
  647741,
  664157,
  647777,
  664161,
  66416

In [238]:
with open("data/terms_dict_by_date.json", "w") as outfile:
    json.dump(terms_dict_by_date, outfile)

In [243]:
#with open('data/terms_dict_by_date.json', 'r') as f:
#    text = json.load(f)